File: sdk/csharp/tests/AnyVali.Tests/DefaultsTests.cs
using AnyVali;
using Xunit;

namespace AnyVali.Tests;

public class DefaultsTests
{
    [Fact]
    public void MissingFieldGetsDefault()
    {
        var schema = V.Object(new Dictionary<string, Schema>
        {
            ["name"] = V.String(),
            ["role"] = V.String().Default("user"),
        });

        var result = schema.SafeParse(new Dictionary<string, object?> { ["name"] = "Alice" });

        Assert.True(result.Success, string.Join(", ", result.Issues.Select(i => i.Code)));
        var data = Assert.IsType<Dictionary<string, object?>>(result.Data);
        Assert.Equal("user", data["role"]);
    }

    [Fact]
    public void PresentFieldIsNotOverwritten()
    {
        var schema = V.Object(new Dictionary<string, Schema>
        {
            ["role"] = V.String().Default("user"),
        });

        var result = schema.SafeParse(new Dictionary<string, object?> { ["role"] = "admin" });

        Assert.True(result.Success);
        var data = Assert.IsType<Dictionary<string, object?>>(result.Data);
        Assert.Equal("admin", data["role"]);
    }

    [Fact]
    public void InvalidDefaultProducesDefaultInvalid()
    {
        var schema = V.Object(new Dictionary<string, Schema>
        {
            ["count"] = V.Int().Min(10).Default(5),
        });

        var result = schema.SafeParse(new Dictionary<string, object?>());

        Assert.False(result.Success);
        Assert.Equal(IssueCodes.DefaultInvalid, result.Issues[0].Code);
        Assert.Equal(new object[] { "count" }, result.Issues[0].Path);
    }

    [Fact]
    public void NullIsNotAbsentForNullableDefault()
    {
        var schema = V.Object(new Dictionary<string, Schema>
        {
            ["value"] = V.Nullable(V.String()).Default("fallback"),
        });

        var result = schema.SafeParse(new Dictionary<string, object?> { ["value"] = null });

        Assert.True(result.Success);
        var data = Assert.IsType<Dictionary<string, object?>>(result.Data);
        Assert.Null(data["value"]);
    }

    [Fact]
    public void FalsyDefaultsAreApplied()
    {
        var schema = V.Object(new Dictionary<string, Schema>
        {
            ["count"] = V.Int().Default(0),
            ["name"] = V.String().Default(""),
            ["active"] = V.Bool().Default(false),
        });

        var result = schema.SafeParse(new Dictionary<string, object?>());

        Assert.True(result.Success, string.Join(", ", result.Issues.Select(i => i.Code)));
        var data = Assert.IsType<Dictionary<string, object?>>(result.Data);
        Assert.Equal(0L, data["count"]);
        Assert.Equal("", data["name"]);
        Assert.Equal(false, data["active"]);
    }

    [Fact]
    public void NestedObjectFieldGetsDefault()
    {
        var schema = V.Object(new Dictionary<string, Schema>
        {
            ["user"] = V.Object(new Dictionary<string, Schema>
            {
                ["name"] = V.String(),
                ["role"] = V.String().Default("guest"),
            }),
        });

        var result = schema.SafeParse(new Dictionary<string, object?>
        {
            ["user"] = new Dictionary<string, object?> { ["name"] = "Bob" },
        });

        Assert.True(result.Success, string.Join(", ", result.Issues.Select(i => i.Code)));
        var data = Assert.IsType<Dictionary<string, object?>>(result.Data);
        var user = Assert.IsType<Dictionary<string, object?>>(data["user"]);
        Assert.Equal("guest", user["role"]);
    }

    [Fact]
    public void MutableDefaultIsNotSharedBetweenParses()
    {
        var schema = V.Object(new Dictionary<string, Schema>
        {
            ["meta"] = V.Any().Default(new Dictionary<string, object?>
            {
                ["items"] = new List<object?>(),
            }),
        });

        var first = Assert.IsType<Dictionary<string, object?>>(
            schema.Parse(new Dictionary<string, object?>()));
        var firstMeta = Assert.IsType<Dictionary<string, object?>>(first["meta"]);
        var firstItems = Assert.IsType<List<object?>>(firstMeta["items"]);
        firstItems.Add("mutated");

        var second = Assert.IsType<Dictionary<string, object?>>(
            schema.Parse(new Dictionary<string, object?>()));
        var secondMeta = Assert.IsType<Dictionary<string, object?>>(second["meta"]);
        Assert.Empty(Assert.IsType<List<object?>>(secondMeta["items"]));
    }
}

File: sdk/rust/tests/defaults_tests.rs
use anyvali::*;
use serde_json::json;

#[test]
fn missing_field_gets_default() {
    let schema = object()
        .field("name", Box::new(string()))
        .field_with_default("role", Box::new(string()), json!("user"))
        .required(vec!["name"]);

    let result = schema.safe_parse(&json!({"name": "Alice"}));

    assert!(result.success, "{:?}", result.issues);
    assert_eq!(result.value.unwrap()["role"], json!("user"));
}

#[test]
fn present_field_is_not_overwritten() {
    let schema = object().field_with_default("role", Box::new(string()), json!("user"));

    let result = schema.safe_parse(&json!({"role": "admin"}));

    assert!(result.success, "{:?}", result.issues);
    assert_eq!(result.value.unwrap()["role"], json!("admin"));
}

#[test]
fn invalid_default_produces_default_invalid() {
    let schema = object().field_with_default("count", Box::new(int().min(10.0)), json!(5));

    let result = schema.safe_parse(&json!({}));

    assert!(!result.success);
    assert_eq!(result.issues[0].code, "default_invalid");
    assert_eq!(
        result.issues[0].path,
        vec![PathSegment::Key("count".to_string())]
    );
}

#[test]
fn null_is_not_absent_for_nullable_default() {
    let schema = object().field_with_default(
        "value",
        Box::new(nullable(Box::new(string()))),
        json!("fallback"),
    );

    let result = schema.safe_parse(&json!({"value": null}));

    assert!(result.success, "{:?}", result.issues);
    assert_eq!(result.value.unwrap()["value"], json!(null));
}

#[test]
fn falsy_defaults_are_applied() {
    let schema = object()
        .field_with_default("count", Box::new(int()), json!(0))
        .field_with_default("name", Box::new(string()), json!(""))
        .field_with_default("active", Box::new(bool_()), json!(false));

    let result = schema.safe_parse(&json!({}));

    assert!(result.success, "{:?}", result.issues);
    assert_eq!(
        result.value.unwrap(),
        json!({"count": 0, "name": "", "active": false})
    );
}

#[test]
fn optional_wrapper_field_gets_default() {
    let schema = object().field(
        "host",
        Box::new(optional(Box::new(string())).default(json!("localhost"))),
    );

    let result = schema.safe_parse(&json!({}));

    assert!(result.success, "{:?}", result.issues);
    assert_eq!(result.value.unwrap(), json!({"host": "localhost"}));
}

#[test]
fn optional_wrapper_default_does_not_override_present_field() {
    let schema = object().field(
        "host",
        Box::new(optional(Box::new(string())).default(json!("localhost"))),
    );

    let result = schema.safe_parse(&json!({"host": "example.com"}));

    assert!(result.success, "{:?}", result.issues);
    assert_eq!(result.value.unwrap(), json!({"host": "example.com"}));
}

#[test]
fn optional_wrapper_default_is_validated() {
    let schema = object().field(
        "host",
        Box::new(optional(Box::new(string().min_length(5))).default(json!("hi"))),
    );

    let result = schema.safe_parse(&json!({}));

    assert!(!result.success);
    assert_eq!(result.issues[0].code, "default_invalid");
    assert_eq!(
        result.issues[0].path,
        vec![PathSegment::Key("host".to_string())]
    );
}

#[test]
fn optional_wrapper_default_is_exported() {
    let node = optional(Box::new(string()))
        .default(json!("localhost"))
        .export_node();

    assert_eq!(node["kind"], json!("optional"));
    assert_eq!(node["default"], json!("localhost"));
}

#[test]
fn mutable_optional_wrapper_default_is_not_shared_between_parses() {
    let schema = object().field(
        "meta",
        Box::new(optional(Box::new(any())).default(json!({"items": []}))),
    );

    let mut first = schema.safe_parse(&json!({})).value.unwrap();
    first["meta"]["items"]
        .as_array_mut()
        .unwrap()
        .push(json!("mutated"));

    let second = schema.safe_parse(&json!({})).value.unwrap();
    assert_eq!(second, json!({"meta": {"items": []}}));
}

#[test]
fn nested_object_field_gets_default() {
    let schema = object()
        .field(
            "user",
            Box::new(
                object()
                    .field("name", Box::new(string()))
                    .field_with_default("role", Box::new(string()), json!("guest"))
                    .required(vec!["name"]),
            ),
        )
        .required(vec!["user"]);

    let result = schema.safe_parse(&json!({"user": {"name": "Bob"}}));

    assert!(result.success, "{:?}", result.issues);
    assert_eq!(result.value.unwrap()["user"]["role"], json!("guest"));
}
